compiler-lib/src/debugging.rs

```diff
@@ -63,36 +63,108 @@ use std::{
-        crate::debugging::pause(
-            crate::debugging::Breakpoint {
-                label: $label.to_string(),
-                line: line!(),
-                column: column!(),
-                file: file!(),
-            },
-            $prog.graph_data(&crate::dot::default_label),
-        );
+        if cfg!(feature = "debugger_vscode") {
+            crate::debugging::show(
+                crate::debugging::Breakpoint {
+                    label: $label.to_string(),
+                    line: line!(),
+                    column: column!(),
+                    file: file!(),
+                },
+                $prog.graph_data(&crate::dot::default_label),
+            );
+        } else {
+            crate::debugging::pause(
+                crate::debugging::Breakpoint {
+                    label: $label.to_string(),
+                    line: line!(),
+                    column: column!(),
+                    file: file!(),
+                },
+                $prog.graph_data(&crate::dot::default_label),
+            );
+        }
     }};
     ($label:expr, $prog:expr, $labels:expr) => {{
         use crate::dot::GraphData;
-        crate::debugging::pause(
-            crate::debugging::Breakpoint {
-                label: $label.to_string(),
-                line: line!(),
-                column: column!(),
-                file: file!(),
-            },
-            $prog.graph_data($labels),
-        );
+        if cfg!(feature = "debugger_vscode") {
+            crate::debugging::show(
+                crate::debugging::Breakpoint {
+                    label: $label.to_string(),
+                    line: line!(),
+                    column: column!(),
+                    file: file!(),
+                },
+                $prog.graph_data($labels),
+            );
+        } else {
+            crate::debugging::pause(
+                crate::debugging::Breakpoint {
+                    label: $label.to_string(),
+                    line: line!(),
+                    column: column!(),
+                    file: file!(),
+                },
+                $prog.graph_data($labels),
+            );
+        }
     }};
 }
 
-#[cfg(not(feature = "debugger_gui"))]
+lazy_static::lazy_static! {
+    static ref LAST_PROG: Mutex<DebuggingHistory>
+        = Mutex::new(DebuggingHistory { items: Vec::new() });
+
+}
+
+struct DebuggingHistory {
+    items: Vec<HashMap<String, GraphState>>,
+}
+
+/// A function where you can place a breakpoint in your debugger
+#[allow(unused_variables)]
+pub fn show(breakpoint: Breakpoint, program: HashMap<String, GraphState>) {
+    if program.len() == 0 {
+        // to get these methods generated so that they can be used for debugging
+        get_all_graph_data();
+        get_last_graph_data();
+    }
+    let mut history = LAST_PROG.lock().unwrap();
+    history.items.push(program);
+}
+
+/// use for debugging
+pub fn get_all_graph_data() -> String {
```

Comment:
is this "pub" to surpress the unused warning? allow(dead_code) would be more explicit.

Reply:
This is pub so I can call it from gdb. But I don't know whether this conclusion is true.

Reply:
ok. thanks